# scAPAtrap v0.2.0 (released on 2023/10/11)
# Identification and quantification of alternative polyadenylation sites from single-cell RNA-seq data

## Introduction
Alternative polyadenylation (APA) has been indicated to play an important role in regulating mRNA stability, translation and localization. Diverse scRNA-seq protocols, such as Drop-seq, CEL-seq, and 10x Genomics, utilizing 3' selection/enrichment in library construction, provide opportunities to extend bioinformatic analysis for studying APA at single cell resolution. We proposed a tool called scAPAtrap for the identification and quantification of APA sites in each individual cells by leveraging the resolution and huge abundance of scRNA-seq data generated by various 3' tag-based protocols. scAPAtrap incorporates peak identification and poly(A) read anchoring, which is capable of identifying and pinpointing poly(A) sites even for those with low read coverage. scAPAtrap can also quantify expression levels of all identified APA sites, considering duplicates resulted from both IVT and PCR cycles. 

scAPAtrap mainly consists of six modules. (1) Raw scRNA-seq datasets from 3’ tag-based protocols (e.g., 10x, CEL-seq) were preprocessed for mapping and extracting UMIs. (2) Without using any genome annotation, potential peaks of the whole genome were detected and wide peaks were iteratively splitted into smaller ones. (3) Identified peaks are quantified by counting effective reads in the peaks. (4) Reads with A/T streches were extracted to determine precise locations of poly(A) sites. (5) Poly(A) sites in both genomic and intergenic regions were annotated with rich information according to the latest genome annotation. (6) Differentially expressed poly(A) sites and 3′ UTR lengthening/shortening events were detected to profile APA dynamics among cell types.

![avatar](https://github.com/BMILAB/scAPAtrap/blob/master/imgs/scAPAtrap_pipeline.png)

## Prerequisites
### Tools
* [samtools](http://www.htslib.org/download/)
* [subread](http://subread.sourceforge.net/)
* [umi_tools](https://github.com/CGATOxford/UMI-tools/blob/master/doc/QUICK_START.md)
* [star](https://github.com/alexdobin/STAR)
* [R](https://cloud.r-project.org/) version 3.6.3, or above.

The above tools can be installed using conda.
```
conda install samtools -c bioconda
conda install subread -c bioconda
conda install umi_tools -c bioconda
conda install star -c bioconda
```
### R packages
Please install the following R packages:
``GenomicRanges`` ``GenomicAlignments`` ``dplyr`` ``derfinder`` ``regionR``

## Installing
```
install.packages('devtools')
devtools::install_github("BMILAB/scAPAtrap")
```

Or you may download the package and install locally.
```
devtools::install_local("your_path_of_scAPAtrap-master.zip", build_vignettes = TRUE)
```

When the package is installed,  you can browse the vignette using the following command on the R console.
```
browseVignettes('scAPAtrap')
```

## Application examples

### Identification and quantification of poly(A) sites 
scAPAtrap tutorial: [PDF](https://github.com/BMILAB/scAPAtrap/blob/master/inst/doc/scAPAtrap_tutorial.pdf), [HTML](https://github.com/BMILAB/scAPAtrap/blob/master/inst/doc/scAPAtrap_tutorial.html) describes two ways of running scAPAtrap: one-step and step-by-step. 

You can also browse the vignette using the following command on the R console
```
vignette("scAPAtrap_tutorial", package = "scAPAtrap")
```

### Analyze APA results from scAPAtrap with the movAPA package
This documentation describes how to read an external file of poly(A) sites and analyze it with movAPA. The vignette was obtained from the [movAPA](https://github.com/BMILAB/movAPA) package.
Please refer to the vignette ([PDF](https://github.com/BMILAB/movAPA/blob/master/inst/doc/movAPA_data_input.pdf), [HTML]( https://bmilab.github.io/movAPA/vignettes/movAPA_data_input.html)) for full details. 

You can also browse the vignette using the following command on the R console
```
vignette("movAPA_on_scAPAtrap_results", package = "scAPAtrap")
```

## Recommended strategy for processing scAPAtrap results
scAPAtrap identifies poly(A) sites through two modes: calling peaks based on change points and finding tails based on soft-clipping. Based on our experience, we recommend the following approach to merge the results from the two modes to obtain a more accurate poly(A) site list.
```
library(movAPA)
library(BSgenome.Mmusculus.UCSC.mm10)

#When setting `tails.search='no'`, the path of the final output file(scAPAtrapData.rda)
pafile=paste0(file.path,"/APA.notails/scAPAtrapData.rda")
load(pafile)
#Load the data into the PACdataset object of movAPA
scPACds=createPACdataset(counts=scAPAtrapData$peaks.count, anno=scAPAtrapData$peaks.meta)
scPACds@colData$barcode<-row.names(scPACds@colData)
scPACds@colData<- subset(scPACds@colData, select= barcode)

#When setting `tails.search=''genome'`, the path of the final output file(scAPAtrapData.rda)
pafile=paste0(file.path,"/APA.tails/scAPAtrapData.rda")
load(pafile)
#Load the data into the PACdataset object of movAPA
sub_scPACds=createPACdataset(counts=scAPAtrapData$peaks.count, anno=scAPAtrapData$peaks.meta)
sub_scPACds@colData$barcode<-row.names(sub_scPACds@colData)
sub_scPACds@colData<- subset(sub_scPACds@colData, select= barcode)
 
#Filter pAs by main chrs
chrs=paste0("chr",c(as.character(1:19),"X","Y")) #mouse
scPACds=subsetPACds(scPACds,chrs=chrs,minExprConds = 0,verbose =T)
sub_scPACds=subsetPACds(sub_scPACds,chrs=chrs,minExprConds = 0,verbose =T)

#Remove internal priming
##First, for all pAs, remove those pAs where there six continuous As within 140 nt upstream and 10 nt downstream

##mouse:BSgenome.Mmusculus.UCSC.mm10/human:BSgenome.Hsapiens.UCSC.hg38
bsgenome=BSgenome.Mmusculus.UCSC.mm10
scPACds =removePACdsIP(scPACds , bsgenome, returnBoth=FALSE, up=-140, dn=10, conA=6, sepA=NA,chrCheck = FALSE)
    
scPACds@anno$PA_id=paste0(scPACds@anno$chr,":",scPACds@anno$strand,":",scPACds@anno$coord)
sub_scPACds@anno$PA_id=paste0(sub_scPACds@anno$chr,":",sub_scPACds@anno$strand,":",sub_scPACds@anno$coord)
    
info=intersect(scPACds@anno$PA_id,sub_scPACds@anno$PA_id)
sub_scPACds@anno=sub_scPACds@anno[sub_scPACds@anno$PA_id %in% info,]
sub_scPACds@counts=sub_scPACds@counts[row.names(sub_scPACds@anno),]

scPACds@anno$pa.tails="no.tail"
loc=match(sub_scPACds@anno$PA_id,scPACds@anno$PA_id)
scPACds@anno$pa.tails[loc]="tail"

##Next, retain the pAs that are supported by the poly(A) tail or have a higher expression level 
##Here, for 3'UTR pAs that are not supported by the poly(A) tail, retain those pAs for which the total expression level is greater than 55% of those 3'UTR pAs that are supported by the poly(A) tail. 

##Note:You can determine which level of expression for the 3'UTR pAs to retain based on the actual conditions

scPACds@anno$count=Matrix::rowSums(scPACds@counts)
    
##Annotate pAs to distinguish 3'UTR pAs
reference.file <- "G:/ref_file/GRcm38fa+gtf/genecode/gencode.vM25.annotation.gtf"#mouse
gff=makeTxDbFromGFF(reference.file , format="gtf")
gff<-parseGenomeAnnotation(gff)    
scPACds<- annotatePAC(scPACds,gff)
scPACds<- ext3UTRPACds(scPACds,ext3UTRlen =2000,extFtr='3UTR')

data=subset(scPACds@anno,scPACds@anno$ftr=="3UTR")
tail.data=subset(data,data$pa.tails=="tail")
#notail.data=subset(data,data$pa.tails=="no.tail") 
    
sub_count=tail.data$count
sub_count=sort(sub_count)
count4=quantile(sub_count,probs=0.25,names=FALSE)
count3=quantile(sub_count,probs=0.50,names=FALSE)
count2=quantile(sub_count,probs=0.55,names=FALSE)
count1=quantile(sub_count,probs=0.60,names=FALSE)

scPACds@anno <-  scPACds@anno %>%
      mutate(count.group = case_when(
        count >count1 ~"count.rank1",
        count > count2 & count <= count1 ~ "count.rank2",
        count > count3 & count <= count2 ~ "count.rank3",
        count > count4 & count <= count3 ~ "count.rank4",
        count <=count4 ~"count.rank5"))
    
scPACds@anno=subset(scPACds@anno,scPACds@anno$pa.tails=="tail" | scPACds@anno$count.group %in% c("count.rank1","count.rank2"))
scPACds@counts=scPACds@counts[row.names(scPACds@anno),]

```

## Analysis in the BiB paper
### Data
Three main datasets used in this study:
* Mouse spermatogenesis ([GSE104556](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE104556))
* Arabidopsis roots ([GSE123013](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE123013))
* Mouse intestinal organoids ([GSE62270](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE62270))

Lists of poly(A) sites with full genome annotation identified by scAPAtrap were placed in the [Result](https://github.com/BMILAB/scAPAtrap/tree/refer/Result) folder.
### Comparions with other tools 
Here we adopted the mouse sperm scRNA-seq dataset to evaluate the performance of scAPAtrap and compared the results with other two tools, scAPA (Shulman et al, 2019) and Sierra (Patrick, et al., 2020). We have used scAPAtrap, scAPA, and Sierra to identify poly(A) sites from the mouse sperm scRNA-seq data, respectively. The identified poly(A) sites stored in Rdata files can be downloaded [here](http://www.bmibig.cn/mnt/scAPAtrap/ScPACdsData/). Please refer to the vignette ([scAPAtrap_compare.html](http://www.bmibig.cn/mnt/scAPAtrap/Tutorial/scAPAtrap_compare.html)) for full details.

### Analysis of APA dynamics
We analyzed dynamic APA usage during sperm cell differentiation based on poly(A) sites identified by scAPAtrap. Please refer to the vignette ([scAPAtrap_DE.html](http://www.bmibig.cn/mnt/scAPAtrap/Tutorial/scAPAtrap_DE.html)) for full details.

## Citation
If you are using scAPAtrap, please cite: [Xiaohui Wu*, Tao Liu, Congting Ye, Wenbin Ye, Guoli Ji: scAPAtrap: identification and quantification of alternative polyadenylation sites from single-cell RNA-seq data, Briefings in Bioinformatics, 2020.](https://pubmed.ncbi.nlm.nih.gov/33142319/)
