---
title: "scAPAtrap: Identification and Quantification of Alternative Polyadenylation Sites from Single-cell RNA-seq Data"
author: 
  - name: Tao Liu
  - name: Xiaohui Wu
    affiliation: Department of Automation, Xiamen University
abstract: "This vignette provides an example on how to use scAPAtrap to detect poly(A) sites from 3'-enriched single-cell RNA sequencing data."
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: 
    BiocStyle::html_document:
    toc: true
    #number_sections: true  ## if you want number sections at each table header
    #theme: united 
keywords: single-cell RNA-seq, alternative polyadenylation, gene regulation, 3' processing
vignette: >
  %\VignetteIndexEntry{scAPAtrap}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup,echo=FALSE}
suppressMessages(library(knitr))
suppressMessages(library(BiocStyle))
suppressMessages(library(GenomicAlignments))
suppressMessages(library(dplyr))
suppressMessages(library(derfinder))
suppressMessages(library(regioneR))
#Color Format
colFmt = function(x,color){
  outputFormat = knitr::opts_knit$get("rmarkdown.pandoc.to")
  if(outputFormat == 'latex')
    paste("\\textcolor{",color,"}{",x,"}",sep="")
  else if(outputFormat == 'html')
    paste("<font color='",color,"'>",x,"</font>",sep="")
  else
    x
}

```
# Introduction
Alternative polyadenylation (APA) has been indicated to play an important role in regulating mRNA stability, translation and localization. Diverse scRNA-seq protocols, such as Drop-seq, CEL-seq, and 10x Genomics, utilizing 3?? selection/enrichment in library construction, provides opportunities to extend bioinformatic analysis for studying APA at single cell resolution. We proposed a tool called scAPAtrap for identification and quantification of APA sites in each individual cells by leveraging the resolution and huge abundance of scRNA-seq data generated by various 3?? tag-based protocols. scAPAtrap incorporates peak identification and poly(A) read anchoring, which is capable of identifying precise locations of APA sites and sites even with low read coverage. scAPAtrap can also quantify the expression levels of all identified APA sites, considering duplicates resulted from both IVT and PCR cycles. 

# scAPAtrap Installation
scAPAtrap can be installed from github.

```
install.packages("devtools")
devtools::install_github("BMILAB/scAPAtrap", build = TRUE, build_vignettes = TRUE)

library(scAPAtrap)
```

Please also install the following tools before running scAPAtrap: 
* [samtools](http://www.htslib.org/download/)
* [subread](http://subread.sourceforge.net/)
* [umi_tools](https://github.com/CGATOxford/UMI-tools/blob/master/doc/QUICK_START.md)
* [star](https://github.com/alexdobin/STAR)

You can also install the above software using conda.

```
conda install samtools -c bioconda
conda install subread -c bioconda
conda install umi_tools -c bioconda
conda install star -c bioconda

```
# DemoData download
Please visit this [link](http://bigbio.xmu.edu.cn/scAPAtrap/demoBam/) to download demo data.The demo data is the data of chromosome 1 in [GSE104556](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE104556). You can also use the entire GSE104556 to run the entire process. Of course, if you do this, you may spend a little more time to understand the entire scAPAtrap process.

# Pipeline
The scAPAtrap toolkit is applicable to majority of poly(A)-captured scRNA-seq protocols. Currently scAPAtrap supports the following library construction technologies: 10x, Drop-seq, InDrops, CEL-seq, CEL-seq2. Due to the different data format of 10x from other protocols, we provide two procedures to show the use of scAPAtrap on 10x and other scRNA-seq protocols (hereinafter called non-10x), respectively.

In this section, we demonstrate how to obtain poly(A) sites from a BAM file of 10x data. For raw fastq file, please refer to the [cellranger tool](https://support.10xgenomics.com/single-cell-gene-expression/software/overview/welcome) on the [official website of 10x](https://www.10xgenomics.com/). For 3'-enriched scRNA-seq technologies other than 10x, cellranger is not applicable to the processing of raw fastq file,Please

First, we configure the following four paths of relevant tools including samtools, umitools, and featureCounts.Note that here you need to replace ``path`` with the yourpath.

```
samtools.path <- 'mypath/bin/samtools'
umitools.path <- 'mypath/bin/umi_tools'
featureCounts.path <- "mypath/bin/featureCounts"
star.path <- 'mypath/bin/STAR'
```
## Preprocessing

### Step 1: findUniqueMap
The BAM file obtained from the 10x tool cellranger normally also contains reads mapped at multiple positions, which would affect the detection especially the quantification of poly(A) sites. Therefore, first we obtain uniquely mapped reads from the input BAM file.

the index parameter and sort parameter in findUniqueMap indicate whether to sort and build the index of the BAM file, which is equivalent to the samtools sort and samtools index commands. It is recommended to sort and build an index here, because in the next steps, the BAM file and its corresponding index file are required

Not all BAM files require the current operation. If the BAM file contains multiple comparison results, such as the BAM file obtained by cellranger, you need to use the findUniqueMap function to perform filtering operations. If not, you do not need to perform this operation.


```
demo.bam <- 'mypath/demo.bam'
nextinput <- findUniqueMap(samtools.path,input = demo.bam,thread = 24,index = T,sort = T)
nextinput
```

### Step 2: dedupByPos
This step removes duplicates and saves only one read per UMI for each position. 

The input of dedupBypos is a BAM file, but the BAM file needs to have a corresponding index. The output is the path where dedupBypos generates the BAM file. The TenX parameter indicates whether the current input BAM file is obtained by the 10x library building technology, if not, please set it to FALSE, such as CEL-seq, Drop-seq.

```
nextinput <- dedupByPos(umitools.path, nextinput, TenX = T)
nextinput
```

### Step 3: separateBamBystrand
The input file of separateBamBystrand is the BAM file generated after the above two steps are processed.
The output will get the corresponding mybam.forward.bam and mybam.reverse.bam respectively representing the BAM file of the reference genome sequence in the forward alignment and the BAM file of the reference genome sequence in the reverse complementary pair alignment.

```
nextinput <- separateBamBystrand(samtools.path, nextinput, 24)
nextinput
```

## findPeaks
After preprocessing, we perform peak calling on each BAM file. Then we will get a file named peaks.saf in the folder "outputdir" which stores the identified peaks.

First, use the ``loadBpCoverages`` function to import the entire BAM file and calculate the coverage for each position of the chromosome.


Then, use ``findPeaks`` to identify the peak according to the calculated coverage, where the maxwidth parameter represents the maximum width of the peak. If the identified peak is greater than this width, continue to divide the peak until the width of the peak is less than 1000bp.

Finally, use generateSAF to generate a SAF file that records the information identifying peak.

```
chrs <- c(as.character(1:19),'X','Y')
maxwidth <- 1000
readlength <- 98
outputdir <- './result'
fullcovF <- loadBpCoverages(nextinput[1],chrs)
fullcovR <- loadBpCoverages(nextinput[2],chrs)

forwardPeaks <-findPeaks(fullcovF,'+',readlength,maxwidth)
reversePeaks <-findPeaks(fullcovR,'-',readlength,maxwidth)

head(forwardPeaks)
head(reversePeaks)

peaksfile <- generateSAF(forwardPeaks, reversePeaks, outputdir)
peaksfile
```

## countPeaks
Based on the peak coordinates obtained from the peak calling step, next we count the read number for each peak using the BAM file and the peak file. This step may take a few minutes depending on the size of the BAM file. Finally, we will get a file named counts.tsv.gz in "outputdir".

```
final.bam <- generateFinalBam(featureCounts.path,samtools.path,demo.bam,peaksfile,24)
final.bam
counts.tsv <- countPeaks(umitools.path,final.bam,outputdir,TenX=T)
counts.tsv
```

## findTails
The peak calling step identifies potential regions where poly(A) sites are located. We can then arbitrarily set the 3' end of each peak as the coordinate of the respective poly(A) site, however, this is not the best solution to determine the precise position of poly(A) site. Here we incorporate reads with A/T stretches to pinpoint poly(A) sites, which was implemented as function findTails in the scAPAtrap package. If the BAM file is large, please use function ``findChrTails`` instead.

```
tails <- findTails(bamfile = demo.bam)
```

## generatescExpMa
scAPAtrap allow for detecting poly(A) sites with the peak calling step and/or the poly(A) read anchoring step. The two steps, although independent, can be jointly considered for detecting poly(A) sites with higher confidence and resolution by taking into account information provided by both poly(A) reads and other reads aligned to the same peak zone. 

We can combine the results of ``findTails`` and ``findPeaks`` to get a refined poly(A) site list. We can also output the poly(A) site data as a PACdataset object which can be used as the input file of the [movAPA package](https://github.com/BMILAB/movAPA) for downstream analysis.

For quantifying poly(A) sites in each cell, scAPAtrap requires a list of valid cell barcodes for the input BAM file. If the BAM file was obtained using CellRanger, then we can simply use the barcodes.tsv file generated in the 'filtered_gene_bc_matrices_mex' folder.

Note, the chromosome naming method in tails must be consistent with the chromosome naming method in peaksfile, for example, either all are chr plus a number, or both are numbers, not one is a number, the other is chr plus a number

```
barcodefile <- './barcode.tsv'
barcode <- read.delim2(barcodefile,header = F)
barcode <- gsub('-[0-9]','',barcode$V1)
expma<- generatescExpMa(countsfile,peaksfile,barcode,tails,min.cells = 2,min.count = 0)

devtools::install_github("BMILAB/movAPA")
coldata <- data.frame(group = colnames(expma)[7:ncol(expma)],row.names = colnames(expma)[7:ncol(expma)])
scPACds <- movAPA::readPACds(pacfile = expma,coldata = coldata)
```

# Percess Fastq
For 3'-enriched scRNA-seq technologies other than 10x, cellranger is not applicable to the processing of raw fastq file. We need some additional preprocessing and then map reads to the genome by read mappers such as STAR.


Please visit this [link](http://bigbio.xmu.edu.cn/scAPAtrap/demoFastq/) to download the demo.fastq file

## Step 1: Extract barcodes and UMIs

Run umi_tools to extract barcodes and UMIs, and add them to read names. For detailed usage of the umi_tools tool, refer to its [help document](https://github.com/CGATOxford/UMI-tools/blob/master/doc/Single_cell_tutorial.md)
```
input.seq <- c('./demo_1.fastq','./demo_2.fastq')
pattern <- 'CCCCCCCCNNNN'
whitelist <- './whitelist.txt'
extractBcAndUb(umitools.path, pattern, input.seq, whitelist)
```

## Step 2: Read mapping

Construction of reference genome index and align.

You can also use the STAR tool to build a reference genome index and alignment on the command line.

```
genome.fasta <- 'mypath/Mus_musculus.GRCm38.dna.primary_assembly.fa'
indexdir <- './mm10_index/'
generateRefIndex(star.path,genome.fasta,indexdir,24)
input.seq <- 'demo_2.extracted.fastq'
generateAlignBam(star.path,indexdir,input.seq,'./BAM/rep1',12)
```

# Session Information-----------------------------------
```{r sessionInfo}
sessionInfo()
```
